reactos/drivers/usb/usbhub_new/pnp.c

```diff
@@ -0,0 +1,2760 @@
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("USBH_StartHubFdoDevice: USBH_GetDeviceType() failed - %lX\n",
+                    Status);
+
+            goto ErrorExit;
+        }
+
+        if (DeviceType == Usb20Device)
+        {
+            HubExtension->HubFlags |= USBHUB_FDO_FLAG_USB20_HUB;
+        }
+    }
+
+    if (HubCount > USBHUB_MAX_CASCADE_LEVELS)
+    {
+        PUSBHUB_PORT_PDO_EXTENSION ParentPdoExtension;
+        PUSBHUB_FDO_EXTENSION ParentHubExtension;
+        USHORT ParentPort;
+        PUSBHUB_PORT_DATA PortData;
+
+        DPRINT1("USBH_StartHubFdoDevice: HubCount > 6 - %x\n", HubCount);
+
+        USBH_WriteFailReasonID(HubExtension->LowerPDO,
+                               USBHUB_FAIL_NESTED_TOO_DEEPLY);
+
+        ParentPdoExtension = HubExtension->LowerPDO->DeviceExtension;
+        ParentHubExtension = ParentPdoExtension->HubExtension;
+
+        ParentPort = ParentPdoExtension->PortNumber - 1;
+        PortData = &ParentHubExtension->PortData[ParentPort];
+        PortData->ConnectionStatus = DeviceHubNestedTooDeeply;
+
+        HubExtension->HubFlags |= USBHUB_FDO_FLAG_DEVICE_FAILED;
+    }
+
+    RtlZeroMemory(&DeviceCapabilities, sizeof(DEVICE_CAPABILITIES));
+
+    USBH_QueryCapabilities(HubExtension->LowerDevice, &DeviceCapabilities);
+
+    HubExtension->SystemWake = DeviceCapabilities.SystemWake;
+    HubExtension->DeviceWake = DeviceCapabilities.DeviceWake;
+
+    RtlCopyMemory(HubExtension->DeviceState,
+                  &DeviceCapabilities.DeviceState,
+                  POWER_SYSTEM_MAXIMUM * sizeof(DEVICE_POWER_STATE));
+
+    Status = USBH_GetDeviceDescriptor(HubExtension->Common.SelfDevice,
+                                      &HubExtension->HubDeviceDescriptor);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("USBH_StartHubFdoDevice: USBH_GetDeviceDescriptor() failed - %lX\n",
+                Status);
+        goto ErrorExit;
+    }
+
+    Status = USBH_GetConfigurationDescriptor(HubExtension->Common.SelfDevice,
+                                             &HubExtension->HubConfigDescriptor);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("USBH_StartHubFdoDevice: USBH_GetConfigurationDescriptor() failed - %lX\n",
+                Status);
+        goto ErrorExit;
+    }
+
+    Status = USBH_SyncGetHubDescriptor(HubExtension);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("USBH_StartHubFdoDevice: USBH_SyncGetHubDescriptor() failed - %lX\n",
+                Status);
+        goto ErrorExit;
+    }
+
+    IsBusPowered = USBH_HubIsBusPowered(HubExtension->Common.SelfDevice,
+                                        HubExtension->HubConfigDescriptor);
+
+    if (IsBusPowered)
+    {
+        /* bus-powered hub is allowed a maximum of 100 mA only for each port */
+        HubExtension->MaxPowerPerPort = 100;
+
+        /* can have 4 ports (4 * 100 mA) and 100 mA remains for itself */
+        HubExtension->HubConfigDescriptor->MaxPower = 250; // 500 mA
```

Comment:
Thus the code needs to be more self-descriptive. I suggest either write a human-readable comment in the code, saying that the spec says that it is expressed in 2 mA units, or, create a macro (called "mA_TO_POWER(...)", for example), that takes in parameter the mA value, and returns the corresponding power value (its implementation would just divide the parameter by 2).
Thus it makes the code readable / understandable, when someone reads it.